### DISCUSSION TITLE
Add a generic context in callbacks

### DISCUSSION THREAD

**Discussion by greg3844** [openssl/openssl#22455] [Q&A]
It's not possible to give a generic pointer (void*) to callback.
I need to use X509_STORE_CTX_verify_cb in a multi-threading process and having a context would be a plus.

Thanks

**Comment by mattcaswell** [comment]
An `X509_STORE_CTX` supports the "ex_data" interface which enables you to associate arbitrary data with the `X509_STORE_CTX` object. Since the `X509_STORE_CTX` gets passed to the verify_cb you can use ex_data to get any application specific context.

See:

https://www.openssl.org/docs/man3.1/man3/X509_STORE_CTX_set_ex_data.html

There is also `X509_STORE_CTX_get_app_data()` and `X509_STORE_CTX_set_app_data()` which presents a simplified interface to this API.

**Comment by greg3844** [comment]
Ah !
Great thanks

**Comment by greg3844** [comment]
I am encountering a problem.
X509_STORE_set_verify_cb use a X509_STORE object but X509_STORE_CTX_set_app_data (like cb fct) use X509_STORE_CTX object.
How can I retrieve  X509_STORE_CTX from X509_STORE ?

**Comment by greg3844** [comment]
I call X509_STORE_CTX_set_app_data() before X509_verify_cert() It's good ?

**Comment by mattcaswell** [comment]
The callback set via `X509_STORE_set_verify_cb` is defined as follows:

````
typedef int (*X509_STORE_CTX_verify_cb)(int, X509_STORE_CTX *);
````
i.e. the callback takes an `X509_STORE_CTX` as an argument, so within the callback you can call `X509_STORE_CTX_get_app_data()`.

The function `X509_verify_cert()` also takes an `X509_STORE_CTX` as an argument, so as long as you have set your app data on the `X509_XTORE_CTX` before calling `X509_verify_cert()` then it will be available to your callback.
